src/iocost/iocost.c

```diff
@@ -0,0 +1,145 @@
+               link);
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",      no_argument,       NULL, 'h'           },
+                { "version",   no_argument,       NULL, ARG_VERSION   },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 1);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+                switch (c) {
+
+                case 'h':
+                        return help();
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        return 1;
+}
+
+static int apply_solution_for_path(const char *path) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_free_ char *qos = NULL, *model = NULL;
+        const char *params;
+        dev_t devnum;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0)
+                return log_error_errno(r, "Error getting devnum for device %s: %m", path);
+
+        r = sd_device_get_property_value(device, "IOCOST_TUNE_MODEL", &params);
+        if (r < 0) {
+                log_info("Parameters for iocost model not available, doing nothing.");
+                return 0;
+        }
+        if (asprintf(&model, DEVNUM_FORMAT_STR " model=linear ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), params) < 0)
+                return log_oom();
+
+        r = sd_device_get_property_value(device, "IOCOST_TUNE_QOS", &params);
+        if (r < 0) {
+                log_info("Parameters for iocost qos not available, doing nothing.");
+                return 0;
+        }
+        if (asprintf(&qos, DEVNUM_FORMAT_STR " enable=1 ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), params) < 0)
+                return log_oom();
+
+        log_debug("Applying iocost parameters to %s\n\tio.cost.model: %s\n\tio.cost.qos: %s\n", path, model, qos);
+
+        r = cg_set_attribute("io", NULL, "io.cost.qos", qos);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set qos: %m");
+
+        r = cg_set_attribute("io", NULL, "io.cost.model", model);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set model: %m");
+
+        return 0;
+}
+
+static int verb_apply(int argc, char *argv[], void *userdata) {
+        const char *path = ASSERT_PTR(argv[1]);
+        return apply_solution_for_path(path);
+}
+
+static int iocost_main(int argc, char *argv[]) {
+        static const Verb verbs[] = {
+                { "apply", 2, 2, 0, verb_apply },
```

Comment:
do you expect to add support for other verbs than "apply"?

we only use the dispatch_verb() stuff when there's benefit of supporting multiple verbs. But if this tool can only do one thing and one thing only anyway, sounds like the wrong way to hook up the command line

Reply:
Yeah, so the original patch had a query verb which is why I had this, and since we are still discussing formats and how things should work I kept the infrastructure around. I have now re-added the query verb as discussed above. Updated branch incoming.